---
title: "nNIL introgression qtl tests"
author: "Jim Holland"
date: "2024-11-21"
output: html_document
---

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(tidyverse)
```

# Get data, filter, and merge BLUEs and introgression calls

Get nNIL BLUEs
```{r}
blues = read.csv("C:/Users/jholland/Box/nNIL genotype data Jim and Tao/GWAS/nNIL_BLUEs.csv")
```

Get introgression summaries per line so we can identify lines with uncertain origin that should be dropped
```{r}
line_summaries = read.csv("C:/Users/jholland/Box/nNIL genotype data Jim and Tao/nNIL_data_supplement/File_S06.nNIL_intro_and_donor_match_summary.csv")

good.lines = line_summaries |>
  filter(all.intros.match.ped) |>
  pull(Line)
```

Keep only nNIL BLUEs if they are in the line_summaries data frame AND if their ped.donor matches best.match donor. This is the set of lines in good.lines object
```{r}
blues_filtered = blues |>
  filter(Geno %in% good.lines)
```


Get introgression calls
```{r}
intros = fread("C:/Users/jholland/Box/nNIL genotype data Jim and Tao/nNIL_data_supplement/File_S19.nNIL_gbs_HMM_introgressionCalls_full_set.csv")
setDF(intros) #convert from data.table to regular data.frame
colnames(intros)[1] = "Geno"
intros_filt = intros |>
  filter(Geno %in% blues_filtered$Geno)
```


Find columns of markers not identical to adjacent marker, discard redundant markers.  
First column is Geno label, 2nd column is first marker, which we always keep.  
Starting at 3rd column of data frame, check if column is NOT identical to preceding column.  
Keep columns that are NOT identical to preceding column.  
any(intros[,x] - intros[,(x-1)]) returns T if any elements are not zero.
```{r}
keepers = unlist(lapply((3:(ncol(intros_filt))), function(x) any(intros_filt[,x] - intros_filt[,(x-1)])))
intros_pruned = intros_filt[,c(T, T, keepers)] #keep first two columns always, Geno column and first marker column, then any non-reduntant columns in keepers
```

Drop any marker columns with no introgressions
```{r}
markers.no.intro = colnames(intros_pruned[-c(1)])[apply(intros_pruned[,-c(1)], 2, function(x) sum(x, na.rm = T) == 0)]
markers.no.intro
```
```{r}
intros_pruned = intros_pruned[,!colnames(intros_pruned) %in% markers.no.intro]
```


Merge filtered BLUEs with pruned markers
```{r}
blues_intros = merge(blues_filtered, intros_pruned, by = "Geno")
```

Get list of marker names
```{r}
markers = colnames(blues_intros)[-c(1:19)]
```

Count number of nNILs with pheno data and passing genotyping filters
```{r}
nrow(blues_intros)
```

# xTreme QTL test

Make a function to get contingency table for one trail and one marker.  
If homoz intros enriched in the tail, then run fisher test.  
Construct a data frame with p.value and counts of nNILs in or not in tail and introgression dose at a marker.
```{r}
tail.introgression.test = function(df, tail, marker){
  #df is the data frame with both phenotypes/tail.occurence and markers for each nNIL
  #tail is the string name of the column to test, eg 'GLS.sig.greater.B73'
  #maker is the string name of the marker column, eg 'S1_53013'
  tab = table(df[,tail], df[,marker], useNA = 'no')

  #check if 1 and 2 scores occur in data set at this marker
  hay1 = "1" %in% dimnames(tab)[[2]]
  hay2 = "2" %in% dimnames(tab)[[2]]
  #test if homozygous introgressions significantly over-represented in tail
  if (hay2) {
    #check if homoz. introgressions over-represented:
    enriched = tab["TRUE", "2"]/tab["FALSE", "2"] > tab["TRUE", "0"]/tab["FALSE", "0"]
    fisher.p = fisher.test(tab)$p.value
    #assemble a data frame with marker name, fisher.p, and the counts of introgression dosage-tail T/F
    if (hay1){ret.df = data.frame(marker = marker, tail = tail, enriched = enriched, fisher.p = fisher.p, F0 = tab["FALSE", "0"], F1 = tab["FALSE", "1"], F2 = tab["FALSE", "2"], T0 = tab["TRUE", "0"], T1 = tab["TRUE", "1"], T2 = tab["TRUE", "2"])} else {ret.df = data.frame(marker = marker, tail = tail, enriched = enriched, fisher.p = fisher.p, F0 = tab["FALSE", "0"], F1 = 0, F2 = tab["FALSE", "2"], T0 = tab["TRUE", "0"], T1 = 0, T2 = tab["TRUE", "2"])}
      return(ret.df)
    } 
  return(data.frame(marker = marker, tail = tail, fisher.p = NA, F0 = NA, F1 = NA, F2 = NA, T0 = NA, T1 = NA, T2 = NA)) #return mostly empty data frame if previous checks not true
}
```

Run the tests for each tail and marker.  
Do NOT include DTA.sig.less.B73 because only 3 genotypes lines in that tail.  
Other tails have > 20 individuals at least.
```{r}
tail.intro.tests = list()
for (t in c("GLS.sig.less.B73", "GLS.sig.greater.B73", "NLB.sig.less.B73", "NLB.sig.greater.B73", "SLB.sig.less.B73", "SLB.sig.greater.B73", "DTA.sig.greater.B73", "EHT.sig.less.B73", "EHT.sig.greater.B73", "PHT.sig.less.B73", "PHT.sig.greater.B73")){
  for (m in colnames(blues_intros)[-c(1:19)]){
    tail.intro.tests[[paste0(t, "_", m)]] = tail.introgression.test(blues_intros, t, m)
  }
}
```

Combine results into a single df
```{r}
results.df = do.call(bind_rows, tail.intro.tests) |>
  mutate(log10p = -log10(fisher.p))
```

Summarize results for each tail
```{r}
results.df |> group_by(tail) |>
  summarize(prop.markers.sign.enriched = sum(enriched & fisher.p < 0.05)/(ncol(intros_pruned)-1))
```

Q-Q plot for all tests by tail
```{r}
ggplot(results.df, aes(sample = log10p)) + geom_qq(aes(fill = tail), distribution = stats::qunif) +
  stat_qq_line(distribution = stats::qunif, colour = "black") +
  ylab("-log10(p)")+
  theme_minimal()
```
To determine appropriate p-value threshold for multiple test correction, we use the formula of Galwey (2009) based on the eigenvalues of the marker correlation matrix.  
The squared sum of sq root eigenvalues divided by the sum of eigenvalues.  
```{r}
corr.intros = cor(intros_pruned[-c(1)])
corr.eig = eigen(corr.intros)
summary(corr.eig$values)
```
Most of the eigenvalues are zero.  These can be ignored for the calculation

```{r}
eig.values.pos = corr.eig$values[corr.eig$values > 0]
nm = ((sum(sqrt(eig.values.pos)))^2)/sum(eig.values.pos)
bonf = 0.05/nm #bonferroni critical p-value for experiment-wide alpha 0.05 correcting for effective number of tests
```


Filter the returned test values to keep only marker-tail combinations significant at bonferroni threshold
```{r}
sign.results.df = results.df |>
  filter(enriched & fisher.p < bonf)

sign.results.df |> group_by(tail) |>
  summarize(prop.markers.sign.enriched = n()/(ncol(intros_pruned)-1))
```
Write out the table of markers enriched in different tails, significant according to the Bonferroni corrected p-value
```{r}
write.csv(sign.results.df, file = "C:/Users/jholland/Box/nNIL genotype data Jim and Tao/GWAS/Markers_sig_enriched_in_tails.csv", row.names = F, quote = F)
```

Write out the pruned introgression and trait data frame
```{r}
write.csv(blues_intros, file = "C:/Users/jholland/Box/nNIL genotype data Jim and Tao/nNIL_data_supplement/File_S23.nNIL_BLUEs_pruned_introgression_markers.csv", row.names = F, quote = F)
```

# QTL test for common allele effect among donors 

Make a function to fit QTL tests. 
Model 1 fits common effect for all donor alleles at each introgression
```{r}
QTL_test_common_effect = function(df, trait, marker){
  y = df[[trait]]
  x = as.factor(df[[marker]])
  mod = lm(y ~ x)
  mod.sum = summary(mod)$coefficients
  mod.sum = as.data.frame(mod.sum[-c(1),c("Estimate", "Pr(>|t|)"), drop = F])
  colnames(mod.sum)[2] = "p.value"
  mod.sum[["Marker"]] = marker
  mod.sum[["Intro_Dosage"]] = row.names(mod.sum)
  mod.sum[["Trait"]] = trait
  return(mod.sum)
}
```

Run the tests
```{r}
results.common.effect.list = list()
for (t in c("GLS", "NLB", "SLB")){
  for (m in markers){
    results.common.effect.list[[paste0(t, "_", m)]] = QTL_test_common_effect(blues_intros, t, m)
  }
}
```

Combine results
```{r}
qtl.results.common.effect.df = do.call(bind_rows, results.common.effect.list)
```

Check p-value distribution
```{r}
ggplot(qtl.results.common.effect.df, aes(x = p.value)) + geom_histogram(aes(fill = Trait)) +
  facet_wrap(facets = vars(Trait)) +
  theme_minimal()
```
Count significant tests, using Bonferroni-corrected p-value computed above
```{r}
qtl.results.common.effect.df |> group_by(Trait) |>
  summarise(N.sign = sum(p.value < bonf))
```

Plot QTL results from model with common donor effects
```{r}
qtl.common.effect.results.sign.df = qtl.results.common.effect.df |> filter(p.value < bonf) |>
  separate(Marker, into = c("chr", "pos"), sep = "_", remove = F) |>
  mutate(chr = as.numeric(sub("S", "", chr)),
         pos = as.numeric(pos),
         Mbp = pos/1000000,
         log10p = -log10(p.value))
```

```{r}
qtl.common.effect.hits.plot = ggplot(qtl.common.effect.results.sign.df, aes(x = Mbp, y = log10p)) +
  geom_point(aes(colour = Trait)) +
  facet_wrap(facets = vars(chr), scales = "free_y") +
  ylab("-log10(p-value)") +
  ggtitle("nNIL introgressions associated with disease traits\ncommon donor effect model") +
  theme_minimal()
qtl.common.effect.hits.plot 
```


Write out results of common QTL effect tests
```{r}
write.csv(qtl.common.effect.results.sign.df , file = "C:/Users/jholland/Box/nNIL genotype data Jim and Tao/GWAS/nNIL_common_introgression_effect_significant_results.csv", row.names = F, quote = F)
```

# Model selection: QTL test for common donor allele effects

Function for forward regression common effect qtl tests
```{r}
forwardQTL = function(df, trait, marker.df, p.thresh = 0.01){
  #df is data frame of trait data and marker scores
  #trait is character string of trait name
  #marker.df is a data.frame with two columns: Marker with character marker names and p.value which are p.values from initial QTL screen
  minp.marker = marker.df |> filter(p.value < p.thresh) 
  if (nrow(minp.marker) == 0) {return(NULL)}
  minp.marker = minp.marker |> slice_min(order_by = p.value, n = 1, with_ties = F) |> pull(Marker)
  marker.df = marker.df |> filter(Marker != minp.marker)

  mod = lm(as.formula(paste(trait, " ~ 1 + ", minp.marker)), data = df) #base model
  best.mod = mod
  while (min(marker.df$p.value < 0.01)){
    results.list = list()
    for (m in marker.df$Marker){
      new.mod = update(best.mod, as.formula(paste("~ . +", m)))
      if (m %in% row.names(summary(new.mod)$coefficients)){ #if new marker causes singularity it will not be in rownames of summary, so have to check for that
      results.list[[m]] = data.frame(Marker = m, p.value = summary(new.mod)$coefficients[m,4])} 
    }
    marker.df = do.call(rbind, results.list) |> filter(p.value < p.thresh) 
    minp.marker = slice_min(marker.df, order_by = p.value, n = 1, with_ties = F) 
    if (nrow(minp.marker) == 0) {return(best.mod)}
    minp.marker = minp.marker |> pull(Marker)
    best.mod = update(best.mod, as.formula(paste("~ . +", minp.marker)))
    marker.df = marker.df |> filter(Marker != minp.marker)
    #if no more markers to test, return best.mod
    if (nrow(marker.df) == 0) {return(best.mod)}
    }
return(best.mod)
}
```


Forward regression to select subset of markers associated with each trait.  
Filter markers to include only those with p < 0.01 as potential markers to include in model  
```{r}
gls.multimod = forwardQTL(blues_intros, "GLS", qtl.common.effect.results.sign.df[qtl.common.effect.results.sign.df$Trait == "GLS",c("Marker", "p.value")])
nlb.multimod = forwardQTL(blues_intros, "NLB", qtl.common.effect.results.sign.df[qtl.common.effect.results.sign.df$Trait == "NLB",c("Marker", "p.value")])
slb.multimod = forwardQTL(blues_intros, "SLB", qtl.common.effect.results.sign.df[qtl.common.effect.results.sign.df$Trait == "SLB",c("Marker", "p.value")])
```

Summarize final multiple QTL model fits
```{r}
summary(gls.multimod)
```
```{r}
summary(nlb.multimod)
```
```{r}
summary(slb.multimod)
```

# QTL for variable donor allele effects

Here we model heterogeneous allele effects for each introgression depending on the donor. To do this we need to join the BLUEs with donor information. We will use the donor matched by sequence to the introgression, and we have already dropped lines where the best.matches based on sequence is not equal to the pedigree donor.  
Get the information on donors for each line.
```{r}
blues_intros2 = merge(line_summaries[,c("Line", "best.matches")], blues_intros, by.x = "Line", by.y = "Geno") |>
  select(-matches("sig")) |>
  rename(donor = best.matches)
```


Make a 2nd function to test introgression*donor interactions. 
Now we have to treat introgression dosage as a numeric value (regressor) with common intercept at 0.  
Otherwise, if marker value is a factor, this model will fit a separate level for the non-introgressed lines for each donor, which is not what we want.  
Once we have a list of markers-traits of interest, we can get their effect estimates later.
```{r}
Intro_Donor_test = function(df, trait, marker){
  y = df[[trait]]
  a = as.numeric(df[[marker]])
  d = 1 - abs(a - 1)# d is 1 for het and 0 for others
  donor = as.factor(df[['donor']])
  mod = aov(y ~ a:donor + d:donor)
  mod.df = summary(mod)[[1]]
  mod.a = mod.df[grepl("^a:donor|donor:a$", trimws(row.names(mod.df))),]
  
  #check if any d effects in model result
  if (any(grepl("^d:donor|donor:d$", trimws(row.names(mod.df))))){
  mod.d = mod.df[grepl("^d:donor|donor:d$", trimws(row.names(mod.df))),]
  DF_dom = mod.d[1,"Df"]
  pval.dom = mod.d[1,"Pr(>F)"]} else { 
    #handle case when no dominance effects in results
    DF_dom = 0
    pval.dom = NA}
  
  coefs = mod$coefficients
  results.df = cbind(data.frame('Marker' = marker, 'Trait' = trait, 'DF_add' = mod.a[1,"Df"], 'DF_dom' = DF_dom, pval.add = mod.a[1,"Pr(>F)"], pval.dom = pval.dom), t(coefs))
  return(results.df)
}
```
Run it
```{r}
donor.intro.results.list = list()
for (t in c("GLS", "NLB", "SLB")){
  for (m in markers){
    donor.intro.results.list [[paste0(t, "_", m)]] = Intro_Donor_test(blues_intros2, t, m)
  }
}
```

```{r}
qtl.donor.intro.results.df = do.call(bind_rows, donor.intro.results.list)
```

Check p-value distribution
```{r}
ggplot(qtl.donor.intro.results.df, aes(x = pval.add)) + geom_histogram(aes(fill = Trait)) +
  facet_wrap(facets = vars(Trait)) +
  theme_minimal()
```
Display significant QTL hits across the genome
```{r}
qtl.donor.intro.results.sign.df = qtl.donor.intro.results.df |> filter(pval.add < bonf) |>
  separate(Marker, into = c("chr", "pos"), sep = "_", remove = F) |>
  mutate(chr = as.numeric(sub("S", "", chr)),
         pos = as.numeric(pos),
         Mbp = pos/1000000,
         log10p = -log10(pval.add))
```


```{r}
qtl.donor.hits.plot = ggplot(qtl.donor.intro.results.sign.df, aes(x = Mbp, y = log10p)) +
  geom_point(aes(colour = Trait)) +
  facet_wrap(facets = vars(chr), scales = "free_y") +
  ylab("-log10(p-value)") +
  ggtitle("nNIL introgressions associated with disease traits\nmarker-by-donor interaction model") +
  theme_minimal()
qtl.donor.hits.plot  
```
Write out results of donor-by-introgression QTL effect tests
```{r}
write.csv(qtl.donor.intro.results.sign.df , file = "C:/Users/jholland/Box/nNIL genotype data Jim and Tao/GWAS/nNIL_donor_by_introgression_interaction_significant_results.csv", row.names = F, quote = F)
```

# Model selection for QTL with variable donor effects  
Function for forward regression donor*QTL interaction tests
```{r}
forwardQTL.donor = function(df, trait, marker.df){
  #df is data frame of trait data and marker scores
  #trait is character string of trait name
  #marker.df is a data.frame with two columns: Marker with character marker names and p.value which are p.values from initial QTL screen
  minp.marker = marker.df |> slice_min(order_by = p.value, n = 1, with_ties = F) |> pull(Marker)
  marker.df = marker.df |> filter(Marker != minp.marker)
  donor = as.factor(df[['donor']])


  mod = lm(as.formula(paste0(trait, " ~ 1 + ", minp.marker, ":donor")), data = df) #base model
  best.mod = mod
  best.aic = AIC(mod)
  keep.on = T
  while (keep.on){
    results.list = list()
    for (m in marker.df$Marker){
      new.mod = update(best.mod, as.formula(paste0("~ . +", m, ":donor")))
      results.list[[m]] = data.frame(Marker = m, aic = AIC(new.mod))} 
    marker.df = do.call(rbind, results.list) 
    minAIC.marker.df = slice_min(marker.df, order_by = aic, n = 1, with_ties = F) 
    if (nrow(minAIC.marker.df) == 0) {return(best.mod)}
    if (minAIC.marker.df$aic >= best.aic) {return(best.mod)}
    minAIC.marker = minAIC.marker.df |> pull(Marker)
    marker.df = marker.df |> filter(Marker != minAIC.marker)
    best.mod = update(best.mod, as.formula(paste0("~ . + ", minAIC.marker, ":donor")))
    best.aic = minAIC.marker.df |> pull(aic)
    }
return(best.mod)
}
```

Forward regression to select subset of markers associated with each trait.  
Filter markers to include only those with pval.add < 0.01 as potential markers to include in model
```{r}
qtl.donor.intro.results.sign.gls = qtl.donor.intro.results.sign.df |>
  filter(Trait == "GLS") |>
  rename(p.value = pval.add)
gls.donor.multimod = forwardQTL.donor(blues_intros2, "GLS", qtl.donor.intro.results.sign.gls)


qtl.donor.intro.results.sign.nlb = qtl.donor.intro.results.sign.df |>
  filter(Trait == "NLB") |>
  rename(p.value = pval.add)
nlb.donor.multimod = forwardQTL.donor(blues_intros2, "NLB", qtl.donor.intro.results.sign.nlb)


qtl.donor.intro.results.sign.slb = qtl.donor.intro.results.sign.df |>
  filter(Trait == "SLB") |>
  rename(p.value = pval.add)

slb.donor.multimod = forwardQTL.donor(blues_intros2, "SLB", qtl.donor.intro.results.sign.slb )
```

Summarize final multiple QTL model fits
```{r}
summary(gls.donor.multimod)
```
```{r}
summary(nlb.donor.multimod)
```

```{r}
summary(slb.donor.multimod)
```

Combine final multi-QTL models and write to hard drive
```{r}
gls.common = as.data.frame(summary(gls.multimod)$coefficients)
gls.common$Marker = row.names(gls.common)
gls.common$Trait = "GLS"

nlb.common = as.data.frame(summary(nlb.multimod)$coefficients)
nlb.common$Marker = row.names(nlb.common)
nlb.common$Trait = "NLB"

slb.common = as.data.frame(summary(slb.multimod)$coefficients)
slb.common$Marker = row.names(slb.common)
slb.common$Trait = "SLB"

all.multiQTL.common = rbind(gls.common, nlb.common, slb.common)
write.csv(all.multiQTL.common, file = "C:/Users/jholland/Box/nNIL genotype data Jim and Tao/GWAS/Common_Effect_MultiQTL_Results.csv", row.names = F, quote = F)
```

```{r}
gls.donor = as.data.frame(summary(gls.donor.multimod)$coefficients)
gls.donor$Marker = row.names(gls.donor)
gls.donor$Trait = "GLS"

nlb.donor = as.data.frame(summary(nlb.donor.multimod)$coefficients)
nlb.donor$Marker = row.names(nlb.donor)
nlb.donor$Trait = "NLB"

slb.donor = as.data.frame(summary(slb.donor.multimod)$coefficients)
slb.donor$Marker = row.names(slb.donor)
slb.donor$Trait = "SLB"

all.multiQTL.donor = rbind(gls.donor, nlb.donor, slb.donor) |>
  filter(!is.na(Estimate))
write.csv(all.multiQTL.donor, file = "C:/Users/jholland/Box/nNIL genotype data Jim and Tao/GWAS/Donor_Effect_MultiQTL_Results.csv", row.names = F, quote = F)
```

# Summaries and visualization of QTL results


Read in the results of xtremeQTL tests:
```{r}
xtreme = read.csv("C:/Users/jholland/Box/nNIL genotype data Jim and Tao/GWAS/Markers_sig_enriched_in_tails.csv") |>
  mutate(Trait = case_when(
    grepl("GLS", tail) ~ "GLS",
    grepl("NLB", tail) ~ "NLB",
    grepl("SLB", tail) ~ "SLB",
    T ~ "NONE"),
        Effect = ifelse(grepl("greater", tail), "Less resistant", "More resistant")) |>
  separate(marker, into = c("chr", "Pos"), sep = "_", remove = F) |>
  mutate(Chr = as.numeric(sub("S", "", chr)),
         Pos = as.numeric(Pos),
         Mbp = Pos/1000000,          #log10p = -log10(p)
          ) |>
  filter(Chr %in% 1:10 & Trait != "NONE") 
```

Read in the results of common donor QTL effect genome scans. Plot only effects of homozygous introgressions.
```{r}
common.QTL.scan = read.csv("C:/Users/jholland/Box/nNIL genotype data Jim and Tao/GWAS/nNIL_common_introgression_effect_significant_results.csv") |>
  mutate(Effect = ifelse(Estimate > 0, "Less resistant", "More resistant"),
         Chr = chr) |>
  filter(Intro_Dosage == "x2")
```

Read in results of multiQTL model selection for common donor QTL effect tests
```{r}
common.QTL.final = read.csv("C:/Users/jholland/Box/nNIL genotype data Jim and Tao/GWAS/Common_Effect_MultiQTL_Results.csv") 
```

Read in results of donor*QTL interaction scans
```{r}
donor.QTL.scan = read.csv("C:/Users/jholland/Box/nNIL genotype data Jim and Tao/GWAS/nNIL_donor_by_introgression_interaction_significant_results.csv") |>
  mutate(log10p = -log10(pval.add),
         Chr = chr)
```


Read in results of multiQTL model selection for donor*QTL interaction tests
```{r}
donor.QTL.final = read.csv("C:/Users/jholland/Box/nNIL genotype data Jim and Tao/GWAS/Donor_Effect_MultiQTL_Results.csv") |>
  filter(!grepl("Intercept", Marker)) |>
  separate(Marker, into = c("pt1", "pt2"), sep = ":", remove = F) |>
  mutate(Marker = ifelse(grepl("donor", pt1), pt2, pt1),
         Donor = ifelse(grepl("donor", pt1), pt1, pt2),
         Donor = sub("donor", "", Donor)) |>
  select(Trait, Marker, Donor, Effect = Estimate, Std.Err = Std..Error, p.value = Pr...t..)

write.csv(donor.QTL.final, "C:/Users/jholland/Box/nNIL genotype data Jim and Tao/GWAS/Donor_Effect_MultiQTL_Results_Nicer.csv", row.names = F, quote = F)
```


Make a data frame with start/end position of each chromosome to use to set x axis scales
```{r}
genome = data.frame(
  Chr = rep(1:10,2),
  Mbp = c(rep(0, 10), 308.45, 243.68, 238.02, 
          250.33, 226.35, 181.36, 185.81,
          	182.41, 163.00, 152.44),
  Trait = "GLS",
  Effect = "More resistant"
)
```


Plot extreme QTL mapping results across genome
```{r}
xtreme2 = bind_rows(xtreme, genome)

xtreme.plot = ggplot(xtreme2, aes(x = Mbp, y = log10p)) +
  geom_point(aes(colour = Trait, shape = Effect), alpha = 0.6) +
  facet_wrap(facets = vars(Chr), scales = "free_x") +
  ylab("-log10(p-value)") +
  ggtitle("nNIL extreme disease tail tests") +
  theme_minimal() 
xtreme.plot 

```
Plot QTL scan with common donor effect
```{r}
common.QTL.scan2 = bind_rows(common.QTL.scan, genome)
highlight.points = merge(common.QTL.scan, common.QTL.final, by = c("Trait", "Marker"))

commonQTL.plot = ggplot(common.QTL.scan2, aes(x = Mbp, y = log10p)) +
  geom_point(aes(colour = Trait, shape = Effect), alpha = 0.6) +
  facet_wrap(facets = vars(Chr), scales = "free_x") +
  ylab("-log10(p-value)") +
  ggtitle("nNIL QTL tests: common donor effect") +
  theme_minimal() +
  #geom_point(aes(x = highlight.points$Mbp, y = highlight.points$log10p), shape = 1, size = 4)
  geom_point(data = highlight.points, aes(color = Trait), shape = 1, size = 4)
commonQTL.plot
```
Plot QTL scan with specific donor effects
```{r}
donor.QTL.scan2 = bind_rows(donor.QTL.scan, genome)

highlight.points.donor = merge(donor.QTL.scan, donor.QTL.final, by = c("Trait", "Marker")) |>
  distinct(Trait, Marker, Chr, Mbp, log10p)

donorQTL.plot = ggplot(donor.QTL.scan2, aes(x = Mbp, y = log10p)) +
  geom_point(aes(colour = Trait), alpha = 0.6) +
  facet_wrap(facets = vars(Chr), scales = "free_x") +
  ylab("-log10(p-value)") +
  ggtitle("nNIL QTL tests: variable donor effects") +
  theme_minimal() +
  #geom_point(aes(x = highlight.points$Mbp, y = highlight.points$log10p), shape = 1, size = 4)
  geom_point(data = highlight.points.donor, aes(color = Trait), shape = 1, size = 4)
donorQTL.plot
```

Plot the distribution of donor effects at QTL in final model.  
Plot the traits separately because they involve different QTL
```{r}
gls.allele.effect.boxplot = ggplot(donor.QTL.final[donor.QTL.final$Trait == "GLS",], aes(x = Marker, y = Effect)) +
  geom_boxplot() +
  ggtitle("Distribution of donor allele effects at GLS QTL") +
  theme_minimal()
gls.allele.effect.boxplot
```
```{r}
ggsave(gls.allele.effect.boxplot, filename = "C:/Users/jholland/Box/nNIL genotype data Jim and Tao/GWAS/GLS_donor_allele_effect_boxplot.tiff", width = 10, height = 8, units = "in", bg = "White")
```


```{r}
nlb.allele.effect.boxplot = ggplot(donor.QTL.final[donor.QTL.final$Trait == "NLB",], aes(x = Marker, y = Effect)) +
  geom_boxplot() +
  ggtitle("Distribution of donor allele effects at NLB QTL") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 

nlb.allele.effect.boxplot
```
```{r}
ggsave(nlb.allele.effect.boxplot, filename = "C:/Users/jholland/Box/nNIL genotype data Jim and Tao/GWAS/NLB_donor_allele_effect_boxplot.tiff", width = 10, height = 8, units = "in", bg = "White")
```


```{r}
slb.allele.effect.boxplot = ggplot(donor.QTL.final[donor.QTL.final$Trait == "SLB",], aes(x = Marker, y = Effect)) +
  geom_boxplot() +
  ggtitle("Distribution of donor allele effects at SLB QTL") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 

slb.allele.effect.boxplot
```
```{r}
ggsave(slb.allele.effect.boxplot, filename = "C:/Users/jholland/Box/nNIL genotype data Jim and Tao/GWAS/SLB_donor_allele_effect_boxplot.tiff", width = 10, height = 8, units = "in", bg = "White")
```


Combine the three QTL scans into a single plot
```{r}
xtreme3 = xtreme2 |>
  select(-chr) |>
  mutate(Scan = "Extreme")
common.QTL.scan = common.QTL.scan |>
  mutate(Scan = "Common QTL")
donor.QTL.scan = donor.QTL.scan |>
  mutate(Scan = "Variable donor QTL")
all.scans = bind_rows(xtreme3, common.QTL.scan, donor.QTL.scan)

highlight.points2 = highlight.points |>
  mutate(Scan = "Common QTL")
highlight.points.donor2 = highlight.points.donor |>
  mutate(Scan = "Variable donor QTL")
highlight.points.all = bind_rows(highlight.points2, highlight.points.donor2)

all.scan.plot = ggplot(all.scans, aes(x = Mbp, y = log10p)) +
  geom_point(aes(colour = Trait), alpha = 0.6) +
  facet_grid(rows = vars(Chr), cols = vars(Scan)) +
  ylab("-log10(p-value)") +
  ggtitle("nNIL QTL tests") +
  theme_minimal() +
  geom_point(data = highlight.points.all, aes(color = Trait), shape = 1, size = 4)

all.scan.plot
```
```{r}
ggsave(all.scan.plot, , filename = "C:/Users/jholland/Box/nNIL genotype data Jim and Tao/GWAS/All_QTL_scan.tiff", width = 10, height = 8, units = "in", bg = "White")
```

