---
title: "SNP_info_nNILs"
author: "Jim Holland"
date: "2024-06-17"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(tidyverse)
library(vcfR)
base.path = "C:/Users/jholland/Box/nNIL genotype data Jim and Tao/"
```

# Get HapMap Data on founders 

Read in the founder HapMap SNPs in v3
```{r}
founderVCF = read.vcfR(paste0(base.path, "Data/nNIL_founders_consistent_gbsSNPs_v3.vcf.gz"), verbose = FALSE )
```

Get the info in v3 on each SNP, get the original ordering of SNPs in vcf
```{r}
founder_meta = founderVCF@fix
founder_meta = cbind(founder_meta, ord = 1:nrow(founder_meta))
```

Get the set of SNPs overlapping between nNIL gbs markers and HapMap3 markers with consistent REF/ALT alleles (identified in Step 4)
```{r}
consistentSNPs = read.csv(paste0(base.path, "Output/nNIL_gbs_SNPs_congruent_w_HapMap_v3v4v5coords_info.csv"))
```

Match the v3 position info with the v4 position info
```{r}
founder_meta_v4 = merge(founder_meta[,c("CHROM", "POS", "ord")], consistentSNPs, by.x = c("CHROM", "POS"), by.y = c("chrV3", "endV3")) %>%
  mutate(ord = as.numeric(ord)) %>%
  arrange(ord)
```

Verify the sorted V3 order matches the order of SNPs in founder matrix
```{r}
all(founder_meta_v4$POS == founder_meta[,"POS"])
```
Get the founder calls
```{r}
founder_calls =extract.gt(founderVCF, return.alleles = T)
founder_calls[founder_calls == "."] = NA
```

We verified above that the sorting of SNP names in founder_meta_v4 matches the founder VCF, so we can use the V4 SNP names to replace the V3 based row names.
```{r}
row.names(founder_calls) = founder_meta_v4$nameV4
colnames(founder_calls) = sub("282set_", "", colnames(founder_calls))
```

Convert the biallelic calls in founder_calls to single character IUPAC values, this is format used by the gbs_snp_v4 data frame.
```{r}
biallelic_to_IUPAC = function(x){
  if (is.na(x)) {return(x)}
  if (x == "A/A") {return("A")}
  if (x == "C/C") {return("C")}
  if (x == "G/G") {return("G")}
  if (x == "T/T") {return("T")}
  if (x == "A/C" | x == "C/A") {return("M")}
  if (x == "A/G" | x == "G/A") {return("R")}
  if (x == "A/T" | x == "T/A") {return("W")}
  if (x == "C/G" | x == "G/C") {return("S")}
  if (x == "C/T" | x == "T/C") {return("Y")}
  if (x == "G/T" | x == "T/G") {return("K")}
  return(NA) #in case no match
}
```

```{r}
founder_calls_iupac = apply(founder_calls, 1:2, biallelic_to_IUPAC)
```

# nNIL GBS SNP calls and introgression calls

Get nNIL GBS SNPs in original V4 positions uplifted to V3 This is a subset after filtering to < 20% missing data in the vcf.
```{r}
gbs_v3 = read.table(paste0(base.path, "Data/nNIL_gbs_snps_converted_V3.bed"))
names(gbs_v3) = c("chrV3", "startV3", "endV3", "nameV4", "score", "strand")
gbs_v3 = gbs_v3 %>% separate(nameV4, into = c("chrV4", "posV4"), sep = "_", remove = F) %>%
  mutate(chrV4 = as.integer(sub("S", "", chrV4)))%>%
  select(-score, -strand)

```

Check GBS nNIL SNP info directly from hapmap meta data column alleles. Assuming first allele is reference
```{r}
gbs_snp_v4 = 
read.table(paste0(base.path,"Data/raw_snps_bgi_id.hmp.txt"), comment.char = "", header = T)
colnames(gbs_snp_v4)[1:4] = c("nameV4", "alleles", "chrV4", "posV4")
```

Now subset the GBS SNP calls on nNIL lines to the same set of SNPs. Make sure missing values of N are set to NA. Also a few rare values need to be set to missing
```{r}
gbs_snp_v4_consistent = gbs_snp_v4[gbs_snp_v4$nameV4 %in% row.names(founder_calls),]
gbs_snp_v4_consistent[gbs_snp_v4_consistent == "-"] = NA
gbs_snp_v4_consistent[gbs_snp_v4_consistent == "N"] = NA
gbs_snp_v4_consistent[gbs_snp_v4_consistent == "O"] = NA
```

change NIL line names from BGI codes to standard NIL line names
```{r}
name_translator = read.table(paste0(base.path, "Data/bgi_nil_id.txt"), sep = "\t", header = T, fileEncoding = "latin1")
name_translator = name_translator %>% 
  mutate(bgi_id = gsub("-", ".", bgi_id),
    nil_id = sub("B73.NIL", "B73 NIL", nil_id))
name_translator_vec = name_translator$nil_id
names(name_translator_vec) = name_translator$bgi_id
colnames(gbs_snp_v4_consistent)[12:ncol(gbs_snp_v4_consistent)] = name_translator_vec[colnames(gbs_snp_v4_consistent)[12:ncol(gbs_snp_v4_consistent)]]
```


Get the introgression calls on the nNILs
```{r}
intros = read.csv(paste0(base.path, "Output/Individual_Introgressions_gbs.csv"))
```


Keep only NILs whose line names match the gbs line names
```{r}
intros_sub = intros |>
  filter(Line %in% colnames(gbs_snp_v4_consistent))
```

Get the introgression calls
```{r}
intro.calls = fread("C:/Users/jholland/Box/nNIL genotype data Jim and Tao/nNIL_data_supplement/File_S18.nNIL_gbs_HMM_introgressionCalls_full_set.csv")
```

Get the list of nNILs from introgression calls
```{r}
allLines = unlist(intro.calls[,1])
```


# Correct SNP ordering and formats

Finally, get the order of SNPs identical between founder_calls_iupac and gbs_snp_v4_consistent. Since the founder gbs and the nNIL gbs were aligned to different reference genomes, their ordering may vary and we need to make them identical
```{r}
all(rownames(founder_calls_iupac) == gbs_snp_v4_consistent$nameV4)
```
Resort the founder call SNPs to match nNILs. Need to parse the names into chrom and position to do this, then sort by NUMERIC chromosome and position
```{r}
founder_call_snps = strsplit(rownames(founder_calls_iupac), "_")
founder_call_snps = as.data.frame(do.call(rbind, founder_call_snps))
colnames(founder_call_snps) = c("chrV4", "posV4")
founder_call_snps$chrV4 = sub("S", "", founder_call_snps$chrV4)
founder_call_snps$nameV4 = rownames(founder_calls_iupac)
founder_calls_iupac = cbind(founder_call_snps, founder_calls_iupac)
founder_calls_iupac = as.data.frame(founder_calls_iupac) %>%
  mutate(chrV4 = as.numeric(chrV4),
         posV4 = as.numeric(posV4)) %>%
  arrange(chrV4, posV4)
```

Check that the new sorting matches the nNIL gbs marker order exactly
```{r}
all(founder_calls_iupac$nameV4 == gbs_snp_v4_consistent$nameV4)
```
Yay!!

# Compare SNP calls within each introgression block in each nNIL to founders

Make a function to compute the proportion of base pair call matches between nNILs and founders within individual introgressions. 
Input is one row of the intros_sub data frame, defining the region of one introgression in one NIL.  
Output is a list with three components: the vector of match percentages against all NAM founders, the number of SNPs, and the percent homoz. calls in the introgression block.
NOTE THAT THERE ARE EDGE CASES WITH FEW, EVEN ZERO, SNPS IN COMMON BETWEEN SMALLER INTROGRESSION BLOCKS IN NILS AND THE FOUNDERS. SO SOME INTROGRESSIONS MAYE RETURN NA VALUES FOR SOME OR ALL FOUNDER MATCH RATES AND POSSIBLY ZERO MARKERS. Function will not fail, but will return NAs.
```{r}
compare_SNPs_in_intros_nohets = function(x){
  #x is one row of intros_sub data frame, becomes a character vector here
  #extract the column of chip SNP calls for the current NIL along with metadata cols 
  line_SNPcalls = gbs_snp_v4_consistent[,c('nameV4','chrV4', 'posV4', x['Line'])]
  #extract the markers defining the current introgression block
  selectMarkers = line_SNPcalls |> 
    filter((chrV4 == as.numeric(x['chr']))  & (posV4 >= as.numeric(x['pos_leftflank'])) & (posV4 <= as.numeric(x['pos_rightflank'])))
  #subset the current NILs SNP calls to only the current introgression
  line_SNPcalls_in_intro = unlist(selectMarkers[x['Line']])

  #subset the founder SNP calls to the same set of markers, dropping cols 1 - 3 of meta-data
  match_founders = founder_calls_iupac[selectMarkers$nameV4,-c(1:3)]
  #compute the match percentage, ignoring hets/NAs
  compare = apply(match_founders, 2, FUN = function(y) {sum(y == line_SNPcalls_in_intro, na.rm = T)/sum(!is.na(y) & !is.na(line_SNPcalls_in_intro))})
  #also compute the number of SNPs NOT missing (not hets)
  N.homoz = sum(!is.na(line_SNPcalls_in_intro))
  return(list(compare, nrow(selectMarkers), N.homoz))
}
```

Apply the function to every row of intros_sub data.frame
```{r}
#extract all of the nested first components of the list components:
comp.list = apply(intros_sub,1,compare_SNPs_in_intros_nohets)
nNIL_compare = lapply(comp.list, function(x) x[[1]])
nNIL_compare = do.call(rbind, nNIL_compare)
colnames(nNIL_compare) = c(colnames(founder_calls_iupac)[-c(1:3)])
#extract all of the nested 2nd components of the list components:
n.SNPs = lapply(comp.list, function(x) x[[2]])
n.SNPs = do.call(c, n.SNPs)
#extract all of the nested 3rd components of the list components:
N.homoz = lapply(comp.list, function(x) x[[3]])
N.homoz = do.call(c, N.homoz)
intros.info = cbind(intros_sub, nNIL_compare)
intros.info$N.SNPs = n.SNPs
intros.info$N.homoz = N.homoz
intros.info$perc.homoz = intros.info$N.homoz/intros.info$N.SNPs
```


Find best match over 0.8 for each introgression  
```{r}
best.matches = apply(nNIL_compare, 1, function(x) colnames(nNIL_compare)[which.max(x)])
matches.80 = apply(nNIL_compare, 1, function(x) x[which.max(x)] > 0.8)
#matches with missing values return logical(0), convert these to FALSE
matches.80 = unlist(lapply(matches.80, function(x) ifelse(length(x) == 0, F, x)))
best.matches[!matches.80]= NA
intros.info$best.matches = unlist(best.matches)
```


Parse the pedigree donor from the line name, fix up donor names
```{r}
intros.info = intros.info |>
  separate_wider_delim(cols = Line, delim = "/", names = c("ped.donor", "suffix"), cols_remove = F) |>
  mutate(ped.donor = ifelse(ped.donor == "MO17", "Mo17", ped.donor),
         ped.donor = ifelse(ped.donor == "KI3", "Ki3", ped.donor),
         best.matches = ifelse(best.matches == "KI3", "Ki3", best.matches),
         ped.match = ped.donor==best.matches,
         intro.size.Mbp = intro.size/1000000)
```

Check that all ped.donor names match hapmap names
```{r}
unique(intros.info$ped.donor)[!unique(intros.info$ped.donor) %in% intros.info$best.matches]
```

```{r}
unique(intros.info$best.matches)[!unique(intros.info$best.matches) %in% intros.info$ped.donor]
```

How many total introgressions were detected?
```{r}
nrow(intros.info)
```

How many unique nNILs were introgressions detected in?
```{r}
length(unique(intros.info$Line))
```

# Visualize distribution of introgression sizes before any filtering

Summary of introgression sizes for all unfiltered introgressions
```{r}
summary(intros.info$intro.size)
```

What proportion of introgressions are < 10kb in size?
```{r}
intros.info |>
  summarize(prop.small.intros = sum(intro.size < 10000)/n())
```

Check for small introgressions carried by multiple nNILs with good matches to pedigree donor:
```{r}
intros.info |>
  filter(intro.size < 5000) |> 
  arrange(chr, pos_leftflank)
```

```{r}
ggplot(intros.info, aes(x = intro.size.Mbp)) +
  geom_histogram(fill = 'blue') +
  xlab('Introgression size (Mb)') +
  geom_vline(xintercept = mean(intros.info$intro.size.Mbp), color = "red") +
  geom_vline(xintercept = median(intros.info$intro.size.Mbp), color = "red", linetype = 2) +
  theme_minimal()
```
```{r}
ggsave(filename = "C:/Users/jholland/Box/nNIL genotype data Jim and Tao/Output/Histogram of Introgression Sizes No Filtering.tiff", dpi = 300, width = 10, height = 8, units = "in", bg = "white")
```


# Visualize introgression size distributions as proportion of total chromosome lengths before filtering

Summary for all introgressions without filtering
```{r}
summary(intros.info$intro.prop.chr)
```


```{r}
ggplot(intros.info, aes(x = intro.prop.chr)) +
  geom_histogram(fill = 'blue') +
  geom_vline(xintercept = mean(intros.info$intro.prop.chr), color = "red") +
  geom_vline(xintercept = median(intros.info$intro.prop.chr), color = "red", linetype = 2) +
  xlab('Introgression proportion of chromosome') +
  theme_minimal()
```
```{r}
ggsave(filename = "C:/Users/jholland/Box/nNIL genotype data Jim and Tao/Output/Histogram of Introgression Proportions of Chromosomes No Filtering.tiff", dpi = 300, width = 10, height = 8, units = "in", bg = "white")
```

# visualize the proportion of genome introgressed into each nNIL

Summarize over Lines, get the amount of sequence and proportion of genome introgressed for each line:
```{r}
V4.total.size = 2120000000
line.summary = intros.info %>%
  group_by(Line) %>%
  summarize(N.blocks.unfiltered = n(),
            Total.Intro.Sizes.Unfiltered = sum(intro.size),
            Total.Intro.Prop.Unfiltered = Total.Intro.Sizes.Unfiltered/V4.total.size) 

summary(line.summary$Total.Intro.Prop.Unfiltered)
```

```{r}
line.summary %>%
  ggplot(aes(x = Total.Intro.Prop.Unfiltered)) +
  geom_histogram(fill = 'blue') +
  xlab("Proportion of genome introgressed per line") +
  geom_vline(xintercept = mean(unlist(line.summary$Total.Intro.Prop.Unfiltered)), color = "red") +
  geom_vline(xintercept = median(unlist(line.summary$Total.Intro.Prop.Unfiltered)), color = "red", linetype = 2) +
  theme_minimal()
```
  
# Inspect match results and filter for high quality matches

Diagnose the matches to B73. Hypothesis is that these are either small introgression blocks or highly heterozygous blocks:
```{r}
intros.info |>
  mutate(GROUP = case_when(
    is.na(best.matches) ~ "NA",
    best.matches == "B73" ~ "B73",
    best.matches == ped.donor ~ "CORRECT",
    best.matches != ped.donor ~ "WRONG"
  )) |>
  group_by(GROUP) |>
  summarise(N.blocks = n(), min.N = min(N.SNPs), mean.N = mean(N.SNPs), max.N = max(N.SNPs),
          min.perc.homoz = min(perc.homoz, na.rm = T), mean.perc.homoz = mean(perc.homoz, na.rm = T), max.perc.homoz = max(perc.homoz, na.rm = T),
          min.N.homoz = min(N.homoz), mean.N.homoz = mean(N.homoz), max.N.homoz = max(N.homoz))
```
Blocks that match B73 are on average smaller than other blocks.  
To be more certain about the founder matches of introgression blocks, apply the following filter rule:  
N.homoz >= 50
```{r}
min.N.homoz = 50

intros.info.filter = intros.info |>
  filter(N.homoz >= min.N.homoz) 
intros.info.filter |>
  mutate(GROUP = case_when(
    is.na(best.matches) ~ "NA",
    best.matches == "B73" ~ "B73",
    best.matches == ped.donor ~ "CORRECT",
    best.matches != ped.donor ~ "WRONG"
  )) |>
  group_by(GROUP) |>
  summarise(N.blocks = n(), min.N = min(N.SNPs), mean.N = mean(N.SNPs), max.N = max(N.SNPs),
          min.perc.homoz = min(perc.homoz), mean.perc.homoz = mean(perc.homoz), max.perc.homoz = max(perc.homoz),
          min.N.homoz = min(N.homoz), mean.N.homoz = mean(N.homoz), max.N.homoz = max(N.homoz))
```

Still one B73 and some NAs, but we expect some of these are real problems. So go with this level of filtering.  

# Summarize introgression matches to donors

SUmmarize number of introgressions over Line and best.matches after filtering
```{r}
intros.info.filter|> group_by(Line, best.matches) |>
  summarize(N.intros = n())
```

Summarize over nNILs the best matches. Do this both for the original set of all called introgressions and also only for filtered introgressions.
```{r}
mismatch.by.NIL = intros.info |>
  group_by(Line, ped.donor) |>
  summarize(N.blocks.unfiltered = n(), N.mismatch.blocks= sum(! ped.match, na.rm = T), N.NA.blocks = sum(is.na(ped.match)), N.donor.matches = sum(!is.na(unique(best.matches))), best.matches = paste(unique(best.matches), sep = "/", collapse = "/")) |>
  mutate(best.matches = sub("NA/|/NA", "", best.matches),
         all.intros.match.ped = ped.donor == best.matches,
          best.matches = ifelse(best.matches == "NA", NA, best.matches))

mismatch.by.NIL.filtered = intros.info.filter |>
  group_by(Line, ped.donor) |>
  summarize(N.blocks.filtered = n(), N.mismatch.blocks= sum(! ped.match, na.rm = T), N.NA.blocks = sum(is.na(ped.match)), N.donor.matches = sum(!is.na(unique(best.matches))), best.matches = paste(unique(best.matches), sep = "/", collapse = "/")) |>
  mutate(best.matches = sub("NA/|/NA", "", best.matches),
         all.intros.match.ped = ped.donor == best.matches,
         best.matches = ifelse(best.matches == "NA", NA, best.matches))

```
Number of nNILs that had no introgressions passing the minimum confidence filter
```{r}
N.nNILs.no.passing.intros = length(unique(intros.info$Line)) - length(unique(mismatch.by.NIL.filtered$Line))
N.nNILs.no.passing.intros
```

How many NILs have at least one mismatched introgression after filtering?
```{r}
mismatch.by.NIL.filtered |>
  filter(N.mismatch.blocks > 0) |>
  nrow()
```

How many NILs match more than one donor after filtering?
```{r}
mismatch.by.NIL.filtered |>
  filter(N.donor.matches > 1) |>
  nrow()
```
How many NILs have no certain matches (all blocks have NA donor match) after filtering?
```{r}
nNILs.no.certain.match = mismatch.by.NIL.filtered |>
  filter(N.donor.matches == 0 & N.mismatch.blocks == 0) |>
  nrow()
nNILs.no.certain.match
```
The total number of nNILs with no high confidence founder matches is this number plus the number of lines with no introgressions with > 50 homoz. SNPs. PLUS the eight lines that had no introgressions called.
```{r}
nNILs.no.hi.confidence.matches = N.nNILs.no.passing.intros + nNILs.no.certain.match + (888 - nrow(mismatch.by.NIL))
nNILs.no.hi.confidence.matches
```
How many NILs have introgression matching ONLY their donor BEFORE filtering individual introgressions for N.homoz SNPs? (Ignoring NA matches)
```{r}
sum(mismatch.by.NIL$all.intros.match.ped, na.rm = T)
```
How many NILs have introgression matching ONLY their donor AFTER filtering individual introgressions for N.homoz SNPs? (Ignoring NA matches)
```{r}
sum(mismatch.by.NIL.filtered$all.intros.match.ped, na.rm = T)
```

# Visualize introgression size distributions after filtering

Compare individual introgression sizes, for nNILs matching donors or not and for filtering based on N.homoz SNPs
```{r}
good.lines = mismatch.by.NIL.filtered |>
  filter(all.intros.match.ped) |>
  pull(Line)

intros.info = intros.info |>
  mutate(nNIL_good = Line %in% good.lines,
         intro_good = N.homoz >= min.N.homoz)

intros.info |>
  group_by(nNIL_good, intro_good) |>
  summarize(min.size = min(intro.size), mean.size = mean(intro.size), max.size = max(intro.size))
```

Summary of introgression sizes for all introgressions in nNILs with good matches to their pedigree donor:
```{r}
intros.info |>
  filter(nNIL_good) |>
  summarize(min.size = min(intro.size), mean.size = mean(intro.size), max.size = max(intro.size))
```

What proportion of introgressions are < 10kb in size?
```{r}
intros.info |>
  group_by(nNIL_good, intro_good) |>
  summarize(prop.small.intros = sum(intro.size < 10000)/n())

```

What proportion of introgressions are < 10kb in size for all introgressions in nNILs with good matches to their pedigree donor?
```{r}
intros.info |>
  filter(nNIL_good) |>
  summarize(prop.small.intros = sum(intro.size < 10000)/n())

```

Check for small introgressions carried by multiple nNILs with good matches to pedigree donor:
```{r}
intros.info |>
  filter(nNIL_good & intro.size < 5000) |> 
  arrange(chr, pos_leftflank)
```

```{r}
intros.info.for.plots  = intros.info  %>%
  mutate(
         donor_match = ifelse(nNIL_good, "nNIL matches donor", "nNIL mismatch donor"),
         intro_good  = ifelse(intro_good, "50+ SNPs in intro", "< 50 SNPs in intro"))
ggplot(intros.info.for.plots , aes(x = intro.size.Mbp)) +
  geom_histogram(fill = 'blue') +
  facet_grid(rows = vars(donor_match), cols =vars(intro_good))  +
  xlab('Introgression size (Mb)') +
  theme_minimal()
```
  
The donor match does not seem to affect introgression size distribution much.  
The N.homoz. SNP filter has a big effect, obviously that is picking out introgressions that are either small or not.  
A simpler display is to include all introgressions carried by NILs that match their pedigree donor:
```{r}
intros.info.for.plots.good.lines = intros.info.for.plots |>
  filter(nNIL_good)

ggplot(intros.info.for.plots.good.lines , aes(x = intro.size.Mbp)) +
  geom_histogram(fill = 'blue') +
  xlab('Introgression size (Mb)') +
  geom_vline(xintercept = mean(intros.info.for.plots.good.lines$intro.size.Mbp), color = "red") +
  geom_vline(xintercept = median(intros.info.for.plots.good.lines$intro.size.Mbp), color = "red", linetype = 2) +
  theme_minimal()
```
```{r}
#ggsave(filename = "C:/Users/jholland/Box/nNIL genotype data Jim and Tao/Output/Histogram of Introgression Sizes Only Good Founder Matches.tiff", dpi = 300, width = 10, height = 8, units = "in", bg = "white")
```

# Visualize introgression size distributions as proportion of total chromosome lengths

Summary for all introgressions without filtering
```{r}
summary(intros.info$intro.prop.chr)
```
Summary for all introgressions among nNILs with good matches to their pedigree donor
```{r}
intros.info |>
  filter(nNIL_good) |>
  pull(intro.prop.chr) |>
  summary()
```

```{r}
ggplot(intros.info.for.plots.good.lines, aes(x = intro.prop.chr)) +
  geom_histogram(fill = 'blue') +
  geom_vline(xintercept = mean(intros.info.for.plots.good.lines$intro.prop.chr), color = "red") +
  geom_vline(xintercept = median(intros.info.for.plots.good.lines$intro.prop.chr), color = "red", linetype = 2) +
  xlab('Introgression proportion of chromosome') +
  theme_minimal()
```
```{r}
#ggsave(filename = "C:/Users/jholland/Box/nNIL genotype data Jim and Tao/Output/Histogram of Introgression Proportions of Chromosomes.tiff", dpi = 300, width = 10, height = 8, units = "in", bg = "white")
```

# visualize the proportion of genome introgressed into each nNIL

Re-check the distribution of proportion of genome introgressed, considering only the NILs with Correct Matches to donors
```{r}
allLines.df  = data.frame(Line = allLines) |>
  separate_wider_delim(cols = Line, delim = "/", names = c("ped.donor2", "suffix"), cols_remove = F) |>
  select(Line, ped.donor2)

line.summary = line.summary %>%
  full_join(mismatch.by.NIL.filtered, by = "Line") |>
  full_join(allLines.df, by = "Line")
line.summary = line.summary |>
  mutate(
    ped.donor = ifelse(is.na(ped.donor), ped.donor2, ped.donor),
    all.intros.match.ped = ifelse(is.na(all.intros.match.ped), F, all.intros.match.ped)) |>#NA values are F
  select(-ped.donor2)

line.summary %>%
  filter(all.intros.match.ped) |>
  pull(Total.Intro.Prop.Unfiltered) |>
  summary()
```
Distribution of total proportion of genome introgressed separately for nNILs that match or not their pedigree donor
```{r}
line.summary %>%
  group_by(all.intros.match.ped) |>
  summarize(mean.intro.rate = mean(Total.Intro.Prop.Unfiltered, na.rm = T),
            N.blocks = mean(N.blocks.unfiltered,na.rm = T))
```



```{r}
line.summary %>%
  ggplot(aes(x = Total.Intro.Prop.Unfiltered)) +
  geom_histogram(fill = 'blue') +
  facet_grid(rows = vars(all.intros.match.ped)) +
  xlab("Proportion of genome introgressed per line") +
  theme_minimal()
```
  
Display the distribution only for the nNILs that match their pedigree donor
```{r}
line.summary %>%
  filter(all.intros.match.ped) |>
  ggplot(aes(x = Total.Intro.Prop.Unfiltered)) +
  geom_histogram(fill = 'blue') +
  xlab("Proportion of genome introgressed per line") +
  geom_vline(xintercept = mean(unlist(line.summary[line.summary$all.intros.match.ped,'Total.Intro.Prop.Unfiltered'])), color = "red") +
  geom_vline(xintercept = median(unlist(line.summary[line.summary$all.intros.match.ped,'Total.Intro.Prop.Unfiltered'])), color = "red", linetype = 2) +
  theme_minimal()
```

```{r}
ggsave(filename = "C:/Users/jholland/Box/nNIL genotype data Jim and Tao/Output/Histogram of Proportion of Whole Genome Introgressed by Line.tiff", dpi = 300, width = 10, height = 8, units = "in", bg = "white")
```

# Summarize the mismatch rates of nNILs according to their pedigree donor
```{r}
line.summary |>
  group_by(ped.donor) |>
  summarize(prop.mismatch = sum(!all.intros.match.ped)/n(),
            n.mismatch = sum(!all.intros.match.ped),
            n.match = sum(all.intros.match.ped),
            N = n())
```



# Save introgression details and match information to files

Save files with details of individual introgressions and their match probabilities before filtering
```{r}
write.csv(intros.info, file = paste0(base.path, "nNIL_data_supplement/File_S05.Individual_Introgressions_gbs_donor_matches.csv"), row.names = F, quote = F)
```

Save the file with details of line introgression donor matches for all lines based on introgressions filtered for N.homoz
```{r}
line.summary = line.summary |>
  select(Line, ped.donor, best.matches, all.intros.match.ped, N.blocks.unfiltered, Total.Intro.Sizes.Unfiltered, Total.Intro.Prop.Unfiltered, N.blocks.filtered, N.mismatch.blocks, N.NA.blocks, N.donor.matches)
write.csv(line.summary, file = paste0(base.path, "nNIL_data_supplement/File_S06.nNIL_intro_and_donor_match_summary.csv"), row.names = F, quote = F)
```

# Distribution of introgression rates by marker

Convert the 0,1,2 introgression dosage calls to 0,1 where 1 is either het or homozygous for the introgression.  
This is so that we measure introgression blocks at their maximum size (if it's homoz in one section but continues at het at another section, we will measure the longest contiguous introgression block length.) If you don't do this, then there will be complications with measuring nested introgressions (a homozygous block inside a het block.).  
Drop first column, which are the line names

```{r}
intros01 = intro.calls[,!1, with = F]
intros01[intros01 > 0] = 1
```
Merge the original introgression calls (1 score means either het or homozygous) with the founder information and map the introgression coverage over the genome
```{r}
setDF(intros01) #convert from data.table to data.frame
intros01 = intros01 |>
  mutate(Line = unlist(intro.calls[,1]))
intros01.founders = merge(line.summary, intros01, by = "Line") 
```

Compute the number of nNILs with an introgression for each marker. First 11 columns of intros01.founders are meta-data, skip these when computing.
```{r}
intro.coverage.by.marker = apply(intros01.founders[,-c(1:11)], 2, sum)
summary(intro.coverage.by.marker)
```
Summarize the introgression coverage as max number of nNILs with an introgression within 1 Mbp bins
```{r}
intro.coverage.by.marker.df = data.frame(Marker = colnames(intros01.founders)[-c(1:11)], coverage = intro.coverage.by.marker) |>
  separate(Marker, into = c("Chr", "Pos"), sep = "_", remove = F) |>
  mutate(Chr = as.numeric(sub("S", "", Chr)),
         Pos = as.numeric(Pos),
         Mbp = Pos/1000000)

intro.coverage.by.bin = intro.coverage.by.marker.df |>
  mutate(bin = round(Mbp, digits = 0)) |>
  group_by(Chr, bin) |>
  summarize(MaxCoverage = max(coverage, na.rm = T))
```
Plot mean introgression coverage by 1 Mbp window
```{r}
intro.coverage.plot = ggplot(intro.coverage.by.bin , aes(x = bin, y = MaxCoverage)) +
  geom_point() +
  facet_wrap(facets = vars(Chr), scales = "free_x") +
  xlab("Mbp") +
  ylab("Number of nNILs with introgression") +
  ggtitle("Introgression coverage by genome position") +
  theme_minimal()

intro.coverage.plot
```
  
Compare introgression coverage using all nNILs vs using only subset of nNILs with correct donor matches
```{r}
intro.coverage.by.marker.good.lines = apply(intros01.founders[intros01.founders$all.intros.match.ped,-c(1:11)], 2, sum)
summary(intro.coverage.by.marker.good.lines)
```
```{r}
intro.coverage.by.marker.good.lines.df = data.frame(Marker = colnames(intros01.founders)[-c(1:11)], coverage = intro.coverage.by.marker.good.lines) |>
  separate(Marker, into = c("Chr", "Pos"), sep = "_", remove = F) |>
  mutate(Chr = as.numeric(sub("S", "", Chr)),
         Pos = as.numeric(Pos),
         Mbp = Pos/1000000)

intro.coverage.by.bin.good.lines = intro.coverage.by.marker.good.lines.df |>
  mutate(bin = round(Mbp, digits = 0)) |>
  group_by(Chr, bin) |>
  summarize(MaxCoverage = max(coverage, na.rm = T))
```
Plot marker coverage for both all nNILs and only good founder match nNILs on same plot
```{r}
intro.coverage.by.bin$Type = "all nNILs"
intro.coverage.by.bin.good.lines$Type = "nNILs matching\npedigree"
intro.coverage.both = bind_rows(intro.coverage.by.bin, intro.coverage.by.bin.good.lines)


intro.coverage.plot.both = ggplot(intro.coverage.both , aes(x = bin, y = MaxCoverage)) +
  geom_point(aes(colour = Type)) +
  facet_wrap(facets = vars(Chr), scales = "free_x") +
  xlab("Mbp") +
  ylab("Number of nNILs with introgression") +
  ggtitle("Introgression coverage by genome position") +
  scale_colour_manual(values = c("blue", "red")) +
  theme_minimal()

intro.coverage.plot.both
```

```{r}
ggsave(intro.coverage.plot.both,filename = "C:/Users/jholland/Box/nNIL genotype data Jim and Tao/Output/Introgression coverage by genome position.tiff", dpi = 300, width = 10, height = 8, units = "in", bg = "white")
```


